libs/parallelism/execution/include/hpx/execution/algorithms/ensure_started.hpp

```diff
@@ -0,0 +1,377 @@
+//  Copyright (c) 2021 ETH Zurich
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/config.hpp>
+#if defined(HPX_HAVE_CXX17_STD_VARIANT)
+#include <hpx/datastructures/tuple.hpp>
+#include <hpx/execution/algorithms/detail/partial_algorithm.hpp>
+#include <hpx/execution/algorithms/detail/single_result.hpp>
+#include <hpx/execution_base/operation_state.hpp>
+#include <hpx/execution_base/receiver.hpp>
+#include <hpx/execution_base/sender.hpp>
+#include <hpx/functional/bind_front.hpp>
+#include <hpx/functional/invoke_fused.hpp>
+#include <hpx/functional/tag_fallback_invoke.hpp>
+#include <hpx/modules/memory.hpp>
+#include <hpx/synchronization/mutex.hpp>
+#include <hpx/thread_support/atomic_count.hpp>
+#include <hpx/type_support/pack.hpp>
+
+#include <boost/container/small_vector.hpp>
+
+#include <atomic>
+#include <cstddef>
+#include <exception>
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <variant>
+
+namespace hpx { namespace execution { namespace experimental {
+    namespace detail {
+        template <typename R>
+        struct error_visitor
+        {
+            std::decay_t<R> r;
+
+            template <typename E>
+            void operator()(E const& e)
+            {
+                hpx::execution::experimental::set_error(std::move(r), e);
+            }
+        };
+
+        template <typename R>
+        struct value_visitor
+        {
+            std::decay_t<R> r;
+
+            template <typename Ts>
+            void operator()(Ts const& ts)
+            {
+                hpx::util::invoke_fused(
+                    hpx::util::bind_front(
+                        hpx::execution::experimental::set_value, std::move(r)),
```

Comment:
I like how the CPO instances can be directly used as functional versions of the API they represent.